### ISSUE TITLE
Documentation of MicroProfile Metrics

### ISSUE THREAD

**Issue by donbourne** [OpenLiberty/docs#5814]
Epic: https://github.com/OpenLiberty/open-liberty/issues/20018

Content: TBD

Things we should document:
1. How to use microprofile metrics with your own micrometer libs so that you can use a different MeterRegistry
    * How to configure meter registries
2. How to use custom scopes for your metrics / how to query for metrics in custom scopes
3. The requirement that the set of tag names for all instances of a metric with the same metric name be the same
----------------------------------
### Existing pages:
* [Monitoring with Metrics](https://openliberty.io/docs/latest/introduction-monitoring-metrics.html) **(Operations > Metrics)** 
   * **Problem: The `MicroProfile Metrics and the /metrics endpoint` section no longer accurately reflect mpMetrics-5.0:**
     * Section makes reference to how to access the `/metrics` endpoint and any sub-trees (i.e. `/metrics/<scope>` and `/metrics/<scope>/<metric name>`. The way to access scope and metric name is no longer with a URL hierarchy, but with parameter query strings (i.e. `/metrics?scope=<scope>` and `/metrics?scope=<scope>&name=<name>`
     * Makes reference to JSON output which is no longer supported.
   * Proposed solution: Where the information is common, leave as is and then separate to a _mpMetrics-4.0 and earlier_ section and `mpMetrics-5.0` section.

* [Metrics reference list](https://openliberty.io/docs/latest/metrics-list.html) ***(Operations > Metrics > Metrics referene list)***
   * **Problem : The existing table is incompatible for mpMetrics-5.0 as well as the few paragraphs above the table.**
      * Prometheus formatted names:
         * scope is no longer prefixed but (i.e. `base_`), but is appended as a tag (i.e. `mp_scope="base"`)
         * Histogram's formatted output no longer contains `stdv` `mean` and `min` values
         * formatting of Prometheus name will always strictly follow `<formattedName>_<unit>_<additionalSuffixes>`
         * `REST.request` metric updated to a `Timer` metric which has a `Histogram` style output. This affects the description as well and the _Version Introduced_.
         * Units are no longer automatically converted.
         * Base and Vendor runtime metrics (except FT related metrics) now register the metric as a base unit. For example, where `cpu.processCpuTime` will show (nanoseconds) in the name column, it will be (seconds) instead.
      * Unrelated to mpMetrics-5.0: Some prometheus names don't show a quotation (`"`) around the tag value.
   * Proposed solution : Duplicate the table for mpMetrics-5.0 and above with updated format names, units, descriptions (if necessary) and include a paragraph before the tables explaining the need for a separate table.

* [Microservice observability with metrics](https://openliberty.io/docs/latest/microservice-observability-metrics.html) **(Development > Observability)**
  * **Problem: The contents of this page do not accurately reflect the contents of mpMetrics-5.0.**
    * The `Simple timer`, `Meter` and `Concurrent gaug` metric type no longer exist.
    * JSON output no longer exists.
      * Similarly, the `OPTION` REST request no longer exists.
      * The JSON output for `Histogram` is no longer applicable.
    * The paths for accessing `/metric` data for specific scopes have changed. No longer is it `/metrics/<scope>`, but rather `/metrics?scope=<scope>`
    * The use of `displayName` in the annotations no longer exists.
    * `Histogram` no longer tracks the standard deviation, minimum and mean values.
    * `Timer` no longer tracks throughput statistics as described on the page. It also behaves (i.e. tracks and presents data) similar to a `Histogram`.

  * Things to add:
    * Point 2 as separate section.
    * Point 3 as a "avoid trouble" type note in the  "Adding metrics to your applications" section.

  * **Separate Problem: There are two pages underneath this path ("Health checks for microservices" and a page/link for "Guides: Observability"). Given the table of contents name of the page (e.g. "Observability") and the title of the page itself and its contents ("Microservice observability with metrics"). The "Observability" page was most likely intended to be a general landing page or top-level page for Observability topics and the actual contents (i.e. "Microservice observability with metrics") was intended to be an article/page under neath.**
  * Proposed solution: Change the "Observability" page to a proper landing page. Place the "Microservice observability with metrics" page as a sub-level page. Duplicate the page for mpMetrics-5.0.

###  New Pages:
- ~A new page to address Point 2 above.~ Include in the [Microservice observability with metrics](https://openliberty.io/docs/latest/microservice-observability-metrics.html) page
_- Perhaps a "Configuring Micrometer target libraries" under (Operations > Metrics > "Configuring Micrometer target libraries"
- Explains how the Open Liberty runtime is based off of Micrometer providing the ability to configure target Micrometer libraries allowing metrics to be forwarded to monitoring systems where there are available target micrometer libraries. Need to explain the difference between the embedded libraries and libraries provided by the customer using the "libraryRef" attribute the `<mpMetrics>` config element._
**- The Configuring with micrometer could be included with the [Monitoring with Metrics](https://openliberty.io/docs/latest/introduction-monitoring-metrics.html) instead. No need to explain how the runtime is running off SmallRye and Micrometer?**

### other:
* Contribute to the differences documentation, similar to [Differences between MicroProfile 5.0 and 4.1](https://openliberty.io/docs/latest/mp-41-50-diff.html)
  * Seems like @dmuelle  was the sole author of this before??
* Update the mpMetrics-5.0 feature page (e.g. link to the mpMetrics-4.0 page https://openliberty.io/docs/latest/reference/feature/mpMetrics-4.0.html) to include the new `libraryRef` attribute. Also a section that quickly shows how to use it. This would be similar to the proposed new page "Configuring Micrometer target libraries", but as more of a  reference and less explaining the underlying mechanism and possibilities.

**Comment by dmuelle** [comment]
Also need to document any new MP config properties at https://openliberty.io/docs/latest/microprofile-config-properties.html#metrics

**Comment by Channyboy** [comment]
@dmuelle  Draft of potential diff section:
I do recall you mentioning that `====` won't render and that it will be figured out, i've left `====` in this draft as well.
```
[#metrics]
== Differences between MicroProfile Metrics 5.0 and 4.0

The MicroProfile Metrics 5.0 release introduces many changes to the specification and programming model that will impact users migrating from MicroProfile Metrics 4.0.

=== SmallRye implementation of the specification

In previous versions, Open Liberty implemented the specification in its own codebase. However, for MicroProfile Metrics 5.0, Open Liberty consumes SmallRye implementation of the specification to produce the feature:mpMetrics-5.0[display=mpMetrics] 5.0 feature.

=== Metrics that have been removed 

The following metric types have been removed from the MicroProfile Metrics 5.0 release:

- Concurrent gauge
- Simple timer
- Meter

=== Metrics that have been updated

The following metric types have been modified in the MicroProfile Metrics 5.0 release or have additional configuration properties from the SmallRye implementation:

==== Histogram

The percentile precision can be configured using the `mp.metrics.smallrye.histogram.precision` MicroProfile Config property. The value is from 1-5, with a larger value resulting in a higher precision for calculating the percentiles at the cost of greater memory usage. The default value is set to 3.

==== Timer
The timer metric no longer tracks throughput in MicroProfile Metrics 5.0. The API has been updated to reflect this.

The percentile precision can be configured using the `mp.metrics.smallrye.timer.precision` MicroProfile Config property. The value is from 1-5, with a larger value resulting in a higher precision for calculating the percentiles at the cost of greater memory usage. The default value is set to 3.

==== Gauge
Gauges now only track numerical values in MicroProfile Metrics 5.0.

=== REST.Request metric
In earlier releases of MicroProfile Metrics, the `REST.Request` metric was instrumented with a `SimpleTimer` metric type. In MicroProfile Metrics 5.0 the `SimpleTimer` metric type is no longer available and the metric is now instrumented with a `Timer` metric type.

=== Metadata

The metric metadata no longer supports the display name and metric unit attributes. 

=== User-defined Metric Registry scope

Starting in MicroProfile Metrics 5.0, in addition to the default `application` metric registry scope, you can define your own metric registry scope to register metrics to. You can define your own scope by specifying the `scope` parameter for metric annotations or by injecting a `MetricRegistry` with the `RegistryScope` annotation and specifying the `scope` parameter. If you don't define a value for the `scope` parameter in the metric annotation or `RegistryScope` annotation, the default scope is `application`.

The following example uses the `@Counted` metric annotation to define a user-defined scope.
[source,java]
----
@Counted(name="customCounter" description="This counter will preside in a user-defined metric registry", scope="customScope")
public String countingThings() {
    return "1 2 3 4..!";
}
----

The following example injects a `MetricRegistry` for instrumenting metrics with the Java API.
[source,java]
----
@Inject
@RegistryScope(scope="customScope")
MetricRegistry customMetricRegistry;
----

=== Metric tags

In prior releases, tags could be used to instrument multiple metrics with the same name as long as the tags had different values. In MicroProfile Metrics 5.0 this is still the case, but is now more strict. The set of key-value pair of tags used must consist of the same tag names.

==== Scope and application name tags

In prior releases, the  `_app` tag defined through the `mp.metrics.appName` MicroProfile Config property is now designated as `mp_app`.
The metric registry scope is also identified with a `mp_scope` tag.

Both of these tags are reserved and if they are used by the user when instrumenting metrics then an `IllegalArgumentException` will occur.

=== REST endpoints

The MicroProfile Metrics 5.0 specification has redefined the `/metrics` REST endpoint in regards to how the metrics output can be filtered with the scope and metric parameters. In previous releases, the REST endpoint used path parameters for querying metrics of a certain scope and metric name with the following syntax `/metrics/<scope>/<name>`. The MicroProfile Metrics 5.0  `/metrics` REST endpoint will now use query parameters in the following syntax `/metrics?scope=<scope>&name=<metric_name>`.

For example, querying for `base` metrics will use the REST endpoint `/metrics?scope=base`. Querying for the `jvm.uptime` metric in the `base scope will use the  REST endpoint `/metrics?scope=base&name=jvm.uptime`.

=== Metric output format

The MicroProfile Metrics 5.0 specification has removed explicit requirement for the JSON output format to be made available. The SmallRye implementation that the Open Liberty runtime is using has also removed JSON output from the `/metrics` endpoint. The specification now only requires that Prometheus formatted metrics be made available through the `/metrics` endpoint.

For the Prometheus formatted output, there have been extensive changes.

==== Prometheus output format of metric registry scope 

The Metric Registry scope that a metric is associated to was indicated with a prefix of `base_` or `vendor_` in the Prometheus formatted name. In MicroProfile Metrics 5.0 this information is displayed with a `mp_scope` tag (i.e. `{mp_scope="base"}`).

==== Prometheus output format of histogram metric

Metrics which were instrumented with a `Histogram` metric type will now have their formatted outputs reduced with the removal of the `min`, `mean` and `stddev` values. Furthermore, the formatting of the Prometheus output has changed. The suffix that defines the `Histogram` value will trail after the unit suffix. For example `<metric_name>_max_seconds` will turn into `<metric_name>_seconds_max`.

==== Prometheus output format of timer metric

Timer metrics will have an output similar to a histogram.

==== Promethues output format of units

In earlier releases, when a metric is instrumented and a unit is defined for it, the Prometheus output will be scaled to the base unit.

In MicroProfile Metrics 5.0, the unit associated to the metric is what is reflected in the Prometheus output. There is no scaling.

=== Integration with Micrometer

In the feature:mpMetrics-5.0[display=mpMetrics] 5.0 feature users are now able to configure Micrometer libraries with Open Liberty during runtime to leverage the Micrometer technology to send metric data to one of 18 supported third-party metric monitoring systems.  This is not a functionality provided by MicroProfile Metrics, but rather with Open Liberty's usage of the SmallRye implementation. See [link to the intoduction-monitoring-metrics.adoc `Integration with Micrometer` section????].
```

**Comment by donbourne** [comment]
preside -> reside

The set of key-value pair of tags used must consist of the same tag names. -> The set of tag names used must be identical across all metric IDs with the same metric name.

In the feature:mpMetrics-5.0[display=mpMetrics] 5.0 feature users are now able to configure Micrometer libraries with Open Liberty during runtime to leverage the Micrometer technology to send metric data to one of 18 supported third-party metric monitoring systems. This is not a functionality provided by MicroProfile Metrics, but rather with Open Liberty's usage of the SmallRye implementation. See [link to the intoduction-monitoring-metrics.adoc Integration with Micrometer section????].
->
The SmallRye metrics implementation, on which Open Liberty bases its feature:mpMetrics-5.0[display=mpMetrics] 5.0 feature, uses Micrometer as its core metrics library. Open Liberty servers can be configured to send metrics to the third-party metric monitoring systems supported by Micrometer. See [link to the intoduction-monitoring-metrics.adoc Integration with Micrometer section????].

**Comment by dmuelle** [comment]
Hi @Channyboy @donbourne thanks for all your help with these drafts. Since we are making significant changes to all the metrics docs, I've been working on reorganizing them a little to make them easier to navigate and use. 

We have metrics docs in both the development and operations section., the idea being the dev info focuses on how to instrument OL/apps to produce metrics and the operations info focusing on how to retrieve those metrics. I've tried to make that separation more consistent, which will hopefully result in less clicking around/searching for users. I know this is somewhat of any arbitrary distinction in places so if there's a better way to organize the docs, just let me know.

I've also tried to reduce some of the redundancy both within and among the various topics. Everything should be sufficiently interlinked so people ca find any of the docs from the one they land on/ start with

Here's a summary of the changes and a few questions:

## [Observability](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/observability.html)

added this short page as a new top-level page in the dev section, since the previous one (Microservice observability with metrics) wasn't really a proper landing page.

## [Microservice observability with metrics](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/microservice-observability-metrics.html)

- focuses on adding metrics to applications
- explanations of tags and scopes, differences for them in 5.0
- rundown of the different annotations with examples and info about differences in 5.0. I collapsed some of the dupklicate sections that had only minor differences for 5.0 and just made note of the differences.
- Some of the examples were duplicated for <4.0 and 5.0, with the only difference being the extra metadata fields in earlier releases. I removed those since the smaller set of metadata is valid in all versions (i think?) and we weren't really explaining anything about those particular fields anyway. IO can restore the 4.0 examples if needed.

## [Choose your own monitoring tools with Micrometer](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/micrometer-metrics.html)

I broke this out into a separate topic because it was making the previous topic even longer, and might have been lost at the bottom of the page

## [Monitoring with metrics](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/introduction-monitoring-metrics.html)

- focuses mainly on the metrics endpoint in 5.09 vs earlier versions, but also includes some JMX info
- not sure the JMX info is helping much, could the little bit that is there just be moved to [Configuring JMX connections](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/configuring-jmx-connection.html)?
- The two tables on this page could probably get by with just the first and last columns? The middle two columns for each table are identical for every row and just restate what is specified in the text introduction to the table.

## [Metrics reference list](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/metrics-list.html)
- tables for 5.0 and 4.0 and earlier. I moved the 5.0 table to be first instead of after 4.0
- also includes info about base vs vendor metrics and how to filter metrics. This info could go in the Development topic, except that the values you'd need to configure filters are in the reference tables so maybe makes more sense for it to be on this page.

## [Differences between MicroProfile Metrics 5.0 and 4.0](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/mp-50-60-diff.html#metrics)

summary of differences

## [MicroProfile Config properties: MicroProfile Metrics](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/microprofile-config-properties.html#metrics)

Updated this table with new properties for 5.0. Please let me know if I'm missing any or if I dont have the values right- particularly the **When the runtime reads the property** column.



Let me know if any further edits or reorganizations are needed. When you're satisfied with the updates, you can add the `technical reviewed` label to this issue to sign off. Thanks!

**Comment by dmuelle** [comment]
@Channyboy - just noticed this. In your changes draft above, for metadata, it says

>The metric metadata no longer supports the display name and metric unit attributes. 

But the [spec](https://download.eclipse.org/microprofile/microprofile-metrics-5.0.0/microprofile-metrics-spec-5.0.0.html#_breaking_changes) indicates it was the display name and _type_ that were removed:

```
Removed displayName from constructor
Removed getDisplayname() method
Removed displayName() method
Removed metricType from constructor
Removed getType() method
Removed getTypeRaw() method
```

1. Should the difference statement be

>The metric metadata no longer supports the display name and metric type attributes. 

2. Also, noticed that the Prometheus output example for histogram returns units, but it's listed as type

> \# TYPE donationDistribution_Dollars summary

The [spec](https://download.eclipse.org/microprofile/microprofile-metrics-5.0.0/microprofile-metrics-spec-5.0.0.html#_histogram) shows the same thing though, units listed under TYPE, so maybe that's expected. Just wanted to double check.

**Comment by Channyboy** [comment]
@dmuelle 
1. Sorry, yes it was the metric type that was removed.
2. In this case  TYPE is Promethues specific and not related to the _Metric Types_ from MP Metrics. The TYPE here is indicated at the end `<formatted_metric_name> <prometheus_type>` (i.e. summary, gauge ,counter)

**Comment by donbourne** [comment]
@dmuelle , in https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/observability.html it says "With Micrometer, you can instrument your code in a standardized way to visualize metrics results in the monitoring system of your choice." -- we don't expose the Micrometer API (app developers use the MP Metrics API and under the covers MP Metrics uses Micrometer), so this statement may need to change to avoid confusion.

**Comment by donbourne** [comment]
@dmuelle , in https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/microservice-observability-metrics.html it says "Counter" and "Gauge" are mp metrics 5.0 only.  those two metric types are present in all versions.

**Comment by donbourne** [comment]
@dmuelle , in https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/micrometer-metrics.html ...

The Open Liberty MicroProfile Metrics 5.0 feature integrates with the Micrometer metrics technology
->
The Open Liberty MicroProfile Metrics 5.0 feature uses embedded Micrometer metrics technology

"Micrometer can connect to 18 different third-party metric monitoring systems" -- we should avoid saying the exact number as this changes over time (eg. I think they list 19 currently and some others are under development).

You can integrate Open Liberty with Micrometer by completing the following steps. 
->
You can integrate Open Liberty with any of the monitoring systems that Micrometer supports by completing the following steps.

**Comment by donbourne** [comment]
@dmuelle , in https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/metrics-list.html

MicrProfile -> MicroProfile

**Comment by dmuelle** [comment]
Thanks for reviewing @donbourne - I've updated the files to address these comments and those you made in #6203, as well as some other small edits and fixes.

- [Observability](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/observability.html)
- [Microservice observability with metrics](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/microservice-observability-metrics.html)
- [Choose your own monitoring tools with Micrometer](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/micrometer-metrics.html)
- [Monitoring with metrics](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/introduction-monitoring-metrics.html)
- [Metrics reference list](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/metrics-list.html)
-  [Differences between MicroProfile Metrics 5.0 and 4.0](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/mp-50-60-diff.html#metrics)
- [MicroProfile Config properties: MicroProfile Metrics](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/microprofile-config-properties.html#metrics)

**Comment by donbourne** [comment]
one comment I'd make overall is that the contrasting of MP Metrics 5.0 and other versions, interwoven (in some cases in the same paragraph), may be useful for migrating users, but will be a bit cumbersome for new MP Metrics 5.0 users or those that have already migrated.  Thinking ahead 1-2 years, users would likely rather be able to skip whole paragraphs/sections that are marked as applying to an older version than have to read through the differences when their goal is to understand MP Metrics 5.0.

**Comment by dmuelle** [comment]
I agree- I think the <4.0 content will be distracting as time goes on. Whenever possible, I've put the 5.0 content/examples before the earlier version stuff to make it easier to skip/ignore. There are a few more changes we might make to cut down on that content:

- on [Microservice observability with metrics](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/microservice-observability-metrics.html), remove the JSON formatted histogram response that only applies to earlier versions- it doesn't add much value.
- on [Monitoring with metrics](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/introduction-monitoring-metrics.html) consolidate multiple references to 4.0 behavior to a single reference to the differences page
- same page- remove the 4.0 and earlier table?

**Comment by dmuelle** [comment]
Hi @Channyboy - your updates from #6316 are now on draft:

https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/micrometer-metrics.html

**Comment by dmuelle** [comment]
Hi @Channyboy @donbourne I believe I have addressed all comments here and this issue is ready for final review/ signoff. The following pages were updated:

- [Observability](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/observability.html)
- [Microservice observability with metrics](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/microservice-observability-metrics.html)
- [Choose your own monitoring tools with Micrometer](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/micrometer-metrics.html)
- [Monitoring with metrics](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/introduction-monitoring-metrics.html)
- [Metrics reference list](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/metrics-list.html)
-  [Differences between MicroProfile Metrics 5.0 and 4.0](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/mp-50-60-diff.html#metrics)
- [MicroProfile Config properties: MicroProfile Metrics](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/microprofile-config-properties.html#metrics)

If any further edits are needed, please let me know. If not, you can sign off by adding the `Technical reviewed` label to this issue. Thanks

**Comment by Channyboy** [comment]
@dmuelle 
The [Choose your own monitoring tools with MicroProfile Metrics](https://github.com/OpenLiberty/docs/issues/5814) has a typo with "Theese ". That was probably my fault :(

**Comment by dmuelle** [comment]
@ramkumar-k-9286 - these updates are ready for peer review:

- [Observability](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/observability.html)
- [Microservice observability with metrics](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/microservice-observability-metrics.html)
- [Choose your own monitoring tools with Micrometer](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/micrometer-metrics.html)
- [Monitoring with metrics](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/introduction-monitoring-metrics.html)
- [Metrics reference list](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/metrics-list.html)
-  [Differences between MicroProfile Metrics 5.0 and 4.0](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/mp-50-60-diff.html#metrics)
- [MicroProfile Config properties: MicroProfile Metrics](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/microprofile-config-properties.html#metrics)

**Comment by ramkumar-k-9286** [comment]
Peer Review
---
[Observability](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/observability.html)

The first sentence seemed a bit confusing when I read it. (Rephrase - added an option)
Observable applications externalize the internal status of a system so operations teams identify and understand the root causes of bugs, bottlenecks, and other inefficiencies.
->
By making the internal status of a system visible to the operations teams, observable applications enable them to identify and understand the root causes of bugs, bottlenecks, and other inefficiencies.

Acrolinx suggestion (Not to use expose): Open Liberty uses MicroProfile Metrics to expose metrics that describe the internal state of many Open Liberty components.
->
Open Liberty uses MicroProfile Metrics to **display/show** metrics that describe the internal state of many Open Liberty components.

The Open Liberty MicroProfile Metrics 5.0 feature uses embedded Micrometer metrics technology to transfer metric data to multiple third-party monitoring systems.
->
Should MicroProfile Metrics 5.0 link be provided here?
Also  Should embedded Micrometer metrics -> Is caps needed for **Micrometer**?

With MicroProfile Health, microservices can self-check their health and publish their overall status to a defined endpoint.
->
Should we add a link here for MicroProfile Health? 
Also.. should it not be "With **the** MicroProfile Health **feature**, microservices..... 

MicroProfile Telemetry helps you collect data on the paths that application requests take through services.
->
I'm not sure if MicroProfile Telemetry is a feature or config.. but the way it is mentioned should there be a link as well as (MicroProfile Telemetry feature/config) not sure about this.

For hands-on tutorials on different observability configurations for Open Liberty, check out our [Observability guides](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/guides/#observability).
->
goes to the Observability guides page, but it seems to be blank.

[Microservice observability with metrics](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/microservice-observability-metrics.html)

Acrolinx suggestion (Not to use expose): Open Liberty uses MicroProfile Metrics to expose metrics that describe the internal state of many Open Liberty components.
->
Open Liberty uses the [MicroProfile Metrics](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/reference/feature/mpMetrics.html) feature to **display/show** metrics that describe the internal state of many Open Liberty components.

For MicroProfile Metrics 4.0 and earlier, only runtime-provided scopes are available and path parameters are used to narrow the scope of metric data. The endpoints are `/metrics/base`, `/metrics/application`, and `/metrics/vendor`.
->
For **[MicroProfile Metrics 4.0] (link)** and earlier, only runtime-provided scopes are available and path parameters are used to narrow the scope of metric data. The endpoints are `/metrics/base`, `/metrics/application`, and `/metrics/vendor`.

By default, metric data is emitted in [Prometheus format](https://prometheus.io/docs/instrumenting/exposition_formats/).
->
By default, metric data is emitted in [Prometheus](https://prometheus.io/docs/instrumenting/exposition_formats/) format.

Starting in version 5.0, MicroProfile Metrics [uses embedded Micrometer metrics technology](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/micrometer-metrics.html) so you can send your metrics data to the third-party monitoring tool of your choice.
->
Starting in version 5.0, MicroProfile Metrics uses [embedded Micrometer metrics technology](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/micrometer-metrics.html) so you can send your metrics data to the third-party monitoring tool of your choice.

Define your own scope either by specifying the `scope` parameter for metric annotations or by injecting a `MetricRegistry` with the `RegistryScope` annotation and specifying the scope parameter.
( Should we define `MetricRegistry`?)

For example, when you use counters, you can create multiple counters named `methodCounter` that return the count of the method that is identified by the tag.
->
For example, when you use counters, you can create **multiple counters that are named** `methodCounter` that return the count of the method that is identified by the tag. ( Acrolinx)

However, in MicroProfile Metrics 5.0 and later, the set of tag names used must be identical across all metric IDs with the same metric name. 
->
However, in MicroProfile Metrics 5.0 and later, the set of tag names **that are used** must be identical across all metric IDs with the same metric name.

For more information, see [Differences between MicroProfile Metrics 5.0 and 4.1](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/mp-50-60-diff.html#metrics).
->
The heading where this gets redirected to says - Differences between MicroProfile Metrics 5.0 and **4.0**

[Concurrent gauge](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/microservice-observability-metrics.html#concurrentgauge) (MicroProfile Metrics 2.0-4.0 only)
->
The link is not working.

Maximum, median and mean values
->
Maximum, median, and mean values

The following example illustrates a histogram that is used to store the value of donations. This example provides the administrator with an idea of the distribution of donation amounts:
->
The following example illustrates a histogram that is used to store the value of donations. This example provides the administrator with an idea of the distribution of donation amounts **.**

For this example, the following response is generated from the REST endpoints in Prometheus format:
->
For this example, the following response is generated from the REST endpoints in Prometheus format **.**

In MicroProfile 4.0 and earlier, the following JSON response is also available from the REST endpoints:
->
In MicroProfile 4.0 and earlier, the following JSON response is also available from the REST endpoints **.**

The counter increments monotonically, counting total invocations of the annotated method:
->
The counter increments monotonically, counting total invocations of the annotated method **.**

Use the `@Gauge` annotation to mark a method as a gauge:
->
Use the `@Gauge` annotation to mark a method as a gauge **.**

The concurrent gauge increments when the annotated method is called and decrements when the annotated method returns, counting current invocations of the annotated method:
->
The concurrent gauge increments when the annotated method is called and decrements when the annotated method returns, counting current invocations of the annotated method **.**

The simple timer tracks how frequently the annotated object is started and how long the invocations take to complete:
->
The simple timer tracks how frequently the annotated object is started and how long the invocations take to complete **.**

This type of metric is available [beginning in MicroProfile Metrics 2.3](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/blog/2020/04/09/microprofile-3-3-open-liberty-20004.html#mra).
->
link not working

[Metrics reference list](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/metrics-list.html)

For information about integrating MicroProfile Metrics 5.0 with Micrometer to send metric data to third-party monitoring systems, see [Choose your own monitoring tools with MicroProfile Metrics](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/micrometer-metrics.html).
->
For **more** information about integrating MicroProfile Metrics 5.0 with Micrometer to send metric data to third-party monitoring systems, see [Choose your own monitoring tools with MicroProfile Metrics](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/micrometer-metrics.html).

To disable all vendor metrics, but to keep the REST base metrics configure the server.xml, as shown in the following example: ( replace : with . )

To disable all monitoring components, add the following code to your server.xml file:( replace : with . )

Q : is Bold -> an accessibility concern here as well?
The **Features required** column of the table includes the feature or features that must be enabled to obtain that metric data.
The **Version introduced** column specifies....

[Differences between MicroProfile Metrics 5.0 and 4.0](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/mp-50-60-diff.html#metrics)

The following metric types are modified in the MicroProfile Metrics 5.0 release or have extra configuration properties from the SmallRye implementation: ( replace : with . )

[MicroProfile Config properties: MicroProfile Metrics](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/microprofile-config-properties.html#metrics)

No comments

**Comment by dmuelle** [comment]
Hi @ramkumar-k-9286 thanks for reviewing. All suggestions implemented except

- per our discussion, did not replace colons with periods
-  Is caps needed for Micrometer? yes, it's a product name
- I didn't add any additional links to the Observability Parent topic, as it's really just to list the child topics
- Q : is Bold -> an accessibility concern here as well? No, it's not used as a replacement for heading styles
